extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts

```diff
@@ -0,0 +1,348 @@
+			this.expandCompleteEmitter.fire(response);
+			return false;
+		} else {
+			setTimeout(() => {
+
+				// Running after promise resolution as we need the Ops Studio-side map to have been updated
+				// Intentionally not awaiting or catching errors.
+				// Any failure in startExpansion should be emitted in the expand complete result
+				// We want this to be async and ideally return true before it completes
+				this.startExpansion(session, nodeInfo, isRefresh);
+			}, 10);
+		}
+		return true;
+	}
+
+	private async startExpansion(session: Session, nodeInfo: sqlops.ExpandNodeInfo, isRefresh: boolean = false): Promise<void> {
+		let expandResult: sqlops.ObjectExplorerExpandInfo = {
+			sessionId: session.uri,
+			nodePath: nodeInfo.nodePath,
+			errorMessage: undefined,
+			nodes: []
+		};
+		try {
+			let node = await session.root.findNodeByPath(nodeInfo.nodePath, true);
+			if (node) {
+				expandResult.errorMessage = node.getNodeInfo().errorMessage;
+				let children = await node.getChildren(true);
+				if (children) {
+					expandResult.nodes = children.map(c => c.getNodeInfo());
+				}
+			}
+		} catch (error) {
+			expandResult.errorMessage = utils.getErrorMessage(error);
+		}
+		this.expandCompleteEmitter.fire(expandResult);
+	}
+
+	refreshNode(nodeInfo: sqlops.ExpandNodeInfo): Thenable<boolean> {
+		// TODO #3815 implement properly
+		return this.expandNode(nodeInfo, true);
+	}
+
+	handleSessionClose(closeSessionInfo: sqlops.ObjectExplorerCloseSessionInfo): Thenable<void> {
+		this.sessionMap.delete(closeSessionInfo.sessionId);
+		return Promise.resolve();
+	}
+
+	findNodes(findNodesInfo: sqlops.FindNodesInfo): Thenable<sqlops.ObjectExplorerFindNodesResponse> {
+		// TODO #3814 implement
+		let response: sqlops.ObjectExplorerFindNodesResponse = {
+			nodes: []
+		};
+		return Promise.resolve(response);
+	}
+
+	registerOnExpandCompleted(handler: (response: sqlops.ObjectExplorerExpandInfo) => any): void {
+		this.expandCompleteEmitter.event(handler);
+	}
+
+	notifyNodeChanged(node: TreeNode): void {
+		this.notifyNodeChangesAsync(node);
+	}
+
+	private async notifyNodeChangesAsync(node: TreeNode): Promise<void> {
+		try {
+			let session = this.getSessionForNode(node);
+			if (!session) {
+				this.appContext.apiWrapper.showErrorMessage(localize('sessionNotFound', 'Session for node {0} does not exist', node.nodePathValue));
+			} else {
+				let nodeInfo = node.getNodeInfo();
+				let expandInfo: sqlops.ExpandNodeInfo = {
+					nodePath: nodeInfo.nodePath,
+					sessionId: session.uri
+				};
+				await this.refreshNode(expandInfo);
+			}
+		} catch (err) {
+			outputChannel.appendLine(localize('notifyError', 'Error notifying of node change: {0}', err));
+		}
+	}
+
+	private getSessionForNode(node: TreeNode): Session {
+		let rootNode: DataServicesNode = undefined;
+		while (rootNode === undefined && node !== undefined) {
+			if (node instanceof DataServicesNode) {
```

Comment:
What is it's the RootNode object that's passed in? I think you should also check 
```suggestion
			if (node instanceof DataServicesNode || node instanceof RootNode) {
```

Reply:
Error: D:/ads/azuredatastudio/extensions/mssql/src/objectExplorerNodeProvider/objectExplorerNodeProvider.ts(198,5): Type 'RootNode | DataServicesNode' is not assignable to type 'DataServicesNode'.
  Type 'RootNode' is not assignable to type 'DataServicesNode'.
    Types have separate declarations of a private property 'children'.